# A00 准备工作

调度系统的复现需要在6号实验的基础上进行，故笔者新建了`DispatchSystem`文件夹之后，将`DeployMultiNodes`中的所有文件都全部复制到了`DispatchSystem`文件夹下备用。

# 01 运行`auto.js`

根据复现手册，运行一个名为`auto.js`的脚本可以自动地创建账户。笔者在代码仓库中搜索到很多文件，在经过比较之后，**认为成佳壮前辈的仓库中的那份文件才是文章中所说的能够创建账户的脚本。**打开一个终端，切换到该脚本所在的路径下，执行：

```bash
node auto.js
```

**！不明晰点！报错了？**

解决方法：打开`auto.js`，翻到`node1 = callfile.exec(...)`发现其中的路径和二进制文件的名字都不太对！？于是将该字符串改为下面这样子（其中一串`cd`的目的是将工作目录切换到00节中提到的`DispatchSystem/Node1`）：

```js
"gnome-terminal -e 'bash -c \"cd ~; cd 文档; cd ReproducingBlockchain; cd DispatchSystem; cd Node1; geth1 --datadir ./gethdata --networkid 91036 --port 30303 --rpc --rpcaddr 127.0.0.1 --rpcport 8545 --rpcapi 'personal,net,eth,web3,admin' --rpccorsdomain='*' --ws --wsaddr='localhost' --wsport 8546 --wsorigins='*' --wsapi 'personal,net,eth,web3,admin' --nodiscover --allow-insecure-unlock --dev.period 1 --syncmode='full' console\"'"
```

解决问题之后，会看到一个启动的控制台。

**！不明晰点！没有按照预想的方式执行代码，报了`Unhandled promise rejection`错误。**

解决方法：研究了半天，原来是因为`gnome-terminal -e`这个语法已经被废弃，被当作是错误给throw出来了，导致后续代码无法执行。于是，将`auto.js`其中的这个部分改为这样（注意`accountConfig`的参数变为8、4、4）：

```js
// ...
async function (error, stdout, stderr) {
    if (stderr) {
        console.log(stderr);
        // throw stderr;
    }
    console.log("ok");
    console.log(stdout);
    sleep(1000);
    
    web3 = new Web3(new Web3.providers.HttpProvider("http://localhost:8545"));
    accountConfig(8, 4, 4);
}
```

**！不明晰点！咋又报`Unhandled promise rejection`了？**

解决方法：将`accountConfig`函数注释掉之后，发现报错消失，控制台不崩溃。因此，一定是`accountConfig`函数出现了问题。

在该函数中上下搜索一番，发现有一句这个：

```js
Promise.all(task).then(...)
```

尝试将它改为：

```js
Promise.all(task).then(...).catch((err) => { console.log(err); })
```

再该一改`createAccount`函数：

问题解决，终端中可以看到一个警告、`result1`和`result2`的内容，并且可以在打开的控制台中进行操作了。



# 02 解锁账户、上传地图并挖矿

在打开的控制台中，输入如下的2条指令来解锁所有账户：

```js
eth.accounts.length
// 16
for (i = 0; i < eth.accounts.length; i++) { personal.unlockAccount(eth.accounts[i],"123456",0) }
// true
```

随后开始挖矿：

```js
miner.start()
```

**！不明晰点！上传地图所需的`uploadmpa._cjz.js`在代码仓库中未找到。**

解决方案：前往成佳壮前辈的仓库中获得了该文件。

观察该文件，发现有这么一段内容：

```js
// contract address 
var myContractInstance = MyContract.at("0xbc9eed91b36c78b87dfdf0c2ca7f0c59111cd620");
var account = "0x9c20114120d1d799dbb680564756aa5a3ff1a72b";
```

复现手册中如下描述：

> 修改`myContractInstance`和`account`（改为现有的一个账户就行）的值

故，笔者如下改动：将`myContractInstance`改为上一个实验中部署合约`StoreMap`的地址；将`account`改为`eth.accounts[0]`的值。

修改完成之后，在终端中运行该脚本：

```bash
node uploadmpa._cjz.js
```

终端不断有新输出，直到出现“地图数据上传完成”字样，关闭终端。

# 03 使用自动化测试

复现手册提及，可以使用`vehicle_test.py`来初始化车辆的状态。简单观察代码后，发现需要安装如下Python库：

```
selenium
```

需要在该文件的`import`部分加上一句：

```python
from importlib import reload
```

**注释** 掉这一句：

```python
sys.setdefaultencoding('utf8')
```

又注意到，代码中使用Chrome进行自动化实验，故需要下载Google Chrome和对应版本的[web driver](https://registry.npmmirror.com/binary.html?path=chromedriver)。

完成以上操作之后，使用如下指令启动脚本：

```bash
python3 vehicle_test.py
```

**！不明晰点！报错：selenium没有叫做`find_element_by_id`的属性。**

解决方法：将所有`find_element_by_id(...)`替换为`find_element(by=By.id, value=...)`。另外，也将所有`find_element_by_class_name(...)`换成`find_element(by=By.CLASS_NAME, value=...)`。

做好以上准备之后，尝试先后运行`vehicle_test.py`和`passenger_test.py`：

```bash
python3 vehicle_test.py
python3 passenger_test.py
```

**！不明晰点！遇到这个错误：`selenium.common.exceptions.TimeoutException`**。很明显，是由于网页没有按照期望的方式运行，导致`wait_until`超时，进而引发的该错误。

